### PR TITLE
Add caching for repeated table data calls

### LINKED ISSUE #177: [Feature]: lru caching for repeated table column data access
### What would you like to see added to the NWBInspector?

Not a huge deal, but I notice it might increase performance on various dandisets to reduce the number of repeated data access operations by the table checks (or anything else that extracts the same limited amount of data across multiple checks).

We had previously hypothesized lru caching for this, so adding it to the to-do list.

### Do you have any interest in helping implement the feature?

Yes.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/NeurodataWithoutBorders/nwbinspector/blob/dev/.github/CODE_OF_CONDUCT.rst)
- [X] Have you ensured this feature was not already [requested](https://github.com/neurodatawithoutborders/nwbinspector/issues)?

### PR DESCRIPTION
**Pull request by CodyCBakerPhD**
fix #177 

Many of the checks we have for tables rely on extracting a specifiable amount of data (`nelems`) from the columns to enact some logic upon. This results in the same pattern of data extraction being applied to the same table column potentially multiple times.

This PR adds caching for the input/output of these calls to allow instant retrieval. The default cache 'size' is `128` function items (pairs of input/output values), which I've bumped to `1000` to get some benefit for larger tables or files with many tables. Each of these pairs should be approximately 1.6 KB in size (assuming `float64` values) for a 1.6 MB cache when fully used.

### CODE REVIEW

File: nwbinspector/utils.py [hunk 1]
```diff
@@ -2,18 +2,30 @@
...
 PathType = TypeVar("PathType", str, Path)  # For types that can be either files or folders
 FilePathType = TypeVar("FilePathType", str, Path)
 OptionalListOfStrings = Optional[List[str]]
 
 dict_regex = r"({.+:.+})"
+MAX_CACHE_ITEMS = 1000  # lru_cache default is 128 calls of matching input/output, but might need more to get use here
+
+
+@lru_cache(maxsize=MAX_CACHE_ITEMS)
+def _subsample_data(data: np.ndarray, nelems: Optional[int] = 200):
```
Comment (bendichter) [inline nwbinspector/utils.py]:
```suggestion
def _subsample_data(data: Union[np.ndarray, h5py.Dataset], nelems: Optional[int] = 200):
```

the whole point here is to cache h5py dataset values, right?

File: nwbinspector/checks/tables.py [hunk 2]
```diff
@@ -79,7 +80,11 @@ def check_time_intervals_stop_after_start(time_intervals: TimeIntervals, nelems:
         very long so you don't need to load the entire array into memory. Use None to
         load the entire arrays.
     """
-    if np.any(np.asarray(time_intervals["stop_time"][:nelems]) - np.asarray(time_intervals["start_time"][:nelems]) < 0):
+    if np.any(
+        np.asarray(_subsample_data(data=time_intervals["stop_time"], nelems=nelems))
+        - np.asarray(_subsample_data(data=time_intervals["start_time"], nelems=nelems))
```
Comment (bendichter) [inline nwbinspector/checks/tables.py]:
```suggestion
        _subsample_data(data=time_intervals["stop_time"], nelems=nelems)
        - _subsample_data(data=time_intervals["start_time"], nelems=nelems)
```

do we really need the np.array wrappers?

File: nwbinspector/utils.py [hunk 3]
```diff
@@ -52,9 +64,12 @@ def check_regular_series(series: np.ndarray, tolerance_decimals: int = 9):
     return len(uniq_diff_ts) == 1
 
 
-def is_ascending_series(series: np.ndarray, nelems=None):
+def is_ascending_series(series: Union[VectorData, ArrayLike], nelems=None):
     """General purpose function for determining if a series is monotonic increasing."""
-    return np.all(np.diff(series[:nelems]) > 0)
+    if isinstance(series, VectorData):
```
Comment (bendichter) [inline nwbinspector/utils.py]: I'd rather test for h5py.Dataset.

File: nwbinspector/utils.py (hunk 1 above)
Comment (bendichter) [reply nwbinspector/utils.py]: I also think I'd prefer to take a tuple of slice objects, so we can cache data more generally

Comment (CodyCBakerPhD) [reply nwbinspector/utils.py]: Done

File: nwbinspector/utils.py (hunk 3 above)
Comment (CodyCBakerPhD) [reply nwbinspector/utils.py]: Done

File: nwbinspector/checks/tables.py [hunk 4]
```diff
@@ -111,7 +111,7 @@ def check_column_binary_capability(table: DynamicTable, nelems: int = 200):
             if np.asarray(column.data[0]).itemsize == 1:
                 continue  # already boolean, int8, or uint8
             try:
-                unique_values = np.unique(_subsample_data(data=column, nelems=nelems))
+                unique_values = np.unique(_cache_data_selection(data=column.data, selection=slice(nelems)))
```
Comment (bendichter) [inline nwbinspector/checks/tables.py:114]:
```suggestion
                unique_values = np.unique(_cache_data_selection(data=column.data, selection=slice(0, nelems)))
```
should this be `slice(0, nelems)`?

File: nwbinspector/utils.py [hunk 5]
```diff
@@ -2,18 +2,30 @@
...
 OptionalListOfStrings = Optional[List[str]]
 
 dict_regex = r"({.+:.+})"
+MAX_CACHE_ITEMS = 1000  # lru_cache default is 128 calls of matching input/output, but might need more to get use here
+
+
+@lru_cache(maxsize=MAX_CACHE_ITEMS)
+def _cache_data_selection(data: h5py.Dataset, selection: Union[slice, Tuple[slice]]) -> np.array:
+    """Extract the selection lazily from the data object for efficient caching (most beneficial during streaming)."""
+    return data[selection]
```
Comment (bendichter) [inline nwbinspector/utils.py:31]:
```suggestion
    data = np.asarray(data) if isinstance(data, list) else data
    return data[selection]
```

File: nwbinspector/utils.py [hunk 6]
```diff
@@ -2,18 +2,31 @@
...
 PathType = TypeVar("PathType", str, Path)  # For types that can be either files or folders
 FilePathType = TypeVar("FilePathType", str, Path)
 OptionalListOfStrings = Optional[List[str]]
 
 dict_regex = r"({.+:.+})"
+MAX_CACHE_ITEMS = 1000  # lru_cache default is 128 calls of matching input/output, but might need more to get use here
+
+
+@lru_cache(maxsize=MAX_CACHE_ITEMS)
+def _cache_data_selection(data: h5py.Dataset, selection: Union[slice, Tuple[slice]]) -> np.array:
```
Comment (bendichter) [inline nwbinspector/utils.py]:
```suggestion
def _cache_data_selection(data: Union[h5py.Dataset, ArrayLike], selection: Union[slice, Tuple[slice]]) -> np.array:
```

**Comment by CodyCBakerPhD** [comment]
OK finally got this working - some technical details arose related to the request to enable selections rather than just `nelems` - apparently, Python `slice`s are not hashable (a requirement of the LRU caching which is essentially just a global dictionary that `kwargs` get updated into): https://stackoverflow.com/questions/29980786/why-are-slice-objects-not-hashable-in-python

There is a simple workaround which is just to parse the representation of each slice in the tuple (which it can easily reconstruct afterward), which is hashable.

Also, other common return types from in-memory `VectorData.data`/`TimeSeries.data` such as lists are also not hashable and so the private helper function that decides whether or not to just slice the data directly or pass it through the `h5py.Dataset`-specific LRU cache (currently, the `h5py.Dataset` object is the only one that benefits from this caching, but when streaming that benefit should be worth it whenever the total quantity of data streamed is larger than the default HDF5 cache but the number of data access commands is smaller than the LRU cache size).

**Comment by bendichter** [comment]
what finesse!

**Review by bendichter — APPROVED** [review]

**Pull request merged**